pkg/tools/serializer/serializer.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package serializer provides true FIFO by-ID lock
+package serializer
+
+import (
+	"sync"
+
+	"github.com/edwarnicke/serialize"
+)
+
+var dummyChan = dummyChannel()
+
+// Serializer is a true FIFO by-ID lock:
+// 1. A -> lock 'a' -- A locked 'a'
+// 2. B -> lock 'a' -- B waiting 'a'
+// 3. C -> lock 'b' -- C locked 'b'
+// 4. A -> unlock 'a', D -> lock 'a' -- B locked 'a', D waiting 'a'
+type Serializer struct {
```

Comment:
Could you please add benchmark tests and share the result?

Reply:
Here are the benchmark tests for the `MutexGroup`:
```
Benchmark_MutexGroup_Different
Benchmark_MutexGroup_Different-4             	 3013191	       455 ns/op
Benchmark_MutexGroup_Same
Benchmark_MutexGroup_Same-4                  	 2828256	       381 ns/op
Benchmark_Executor
Benchmark_Executor-4                         	 1269505	      1019 ns/op
Benchmark_MutexGroup_Different_Goroutine
Benchmark_MutexGroup_Different_Goroutine-4   	  494902	      2116 ns/op
Benchmark_MutexGroup_Same_Goroutine
Benchmark_MutexGroup_Same_Goroutine-4        	 1000000	      3666 ns/op
Benchmark_Executor_Goroutine
Benchmark_Executor_Goroutine-4               	 1000000	      1438 ns/op
```